File: IncreasingOrder.kt
fun main(args: Array<String>) {
    println(almostIncreasingSequence(arrayOf(1, 4, 10, 4, 2))) // false
    println(almostIncreasingSequence(arrayOf(1, 2, 1, 2))) // false
    println(almostIncreasingSequence(arrayOf(1, 3, 2))) // true
    println(almostIncreasingSequence(arrayOf(1, 3, 2, 1))) // false
    println(almostIncreasingSequence(arrayOf(1, 3))) // true
    println(almostIncreasingSequence(arrayOf(1))) // true
    println(almostIncreasingSequence(arrayOf())) // false
}

/**
 * Checks if an array of Integers obtains an ascending order if one single element is removed.
 * @param array The array to be checked if removing a single element, it becomes ordered by ascending.
 */
fun almostIncreasingSequence(array: Array<Int>) = if (array.isEmpty()) {
    // If the list is empty, return false, as an empty list can't have an order.
    false
} else {
    // Retrieve the index of an element that breaks the order
    val index = getUnorderedElementIndex(array)
    // If the index is -1 (no element breaks the order), the list is already in ascending order.
    if (index == -1) {
        true
    } else {
        val arrayWithoutPreviousElement = array.copyOfRange(0, index - 1)
                .plus(array.copyOfRange(index, array.size))
        val arrayWithoutNextElement = array.copyOfRange(0, index + 1)
                .plus(array.copyOfRange(index + 2, array.size))
        // If not, check if by removing the previous element at the retrieved index orders the list.
        getUnorderedElementIndex(arrayWithoutPreviousElement) == -1 ||
                // Or check if by removing the next element at the retrieved index orders the list.
                getUnorderedElementIndex(arrayWithoutNextElement) == -1
    }
}

/**
 * Checks if a given array of Integers is currently in an ascending order.
 * @param array array of integers being checked.
 */
fun getUnorderedElementIndex(array: Array<Int>): Int {
    // Iterate through all the elements in the array except for the last one.
    for (i in 0 until array.size - 1) {
        // If the current element is bigger or equals than the next one, return the element index.
        if (array[i] >= array[i + 1]) {
            return i
        }
    }
    // If said condition is never met, return a non index position, a -1 can work.
    return -1
}

File: IncreasingOrder.swift
import Foundation 

/**
* Checks if an array of integers is already in ascending order.
**/
func getUnorderedElementIndex(array: [Int]) -> Int {
    // Iterate through all the elements of the array, except for the last one.
    for index in array.indices.dropLast() {
        // If the current element is equals or larger than the next element
        if(array[index] >= array[index + 1]) {
            // return the current element position
            return index
        }
    }
    // If said condition is not met, return a non-index value.
    return -1
}

/**
* Check if a given array becomes ordered by ascending if a single element
* gets removed.
**/
func almostIncreasingSequence(sequence: [Int]) -> Bool {
    // If the array is empty, there is nothing to check, so return false
    if(sequence.isEmpty) {
        return false
    } else {
        // If the array is not empty, ge tthe index of the element that disrupts the order in the array.
        let index = getUnorderedElementIndex(array: sequence)
        // If the index is -1 (no element disrupts the array), the array is already ordered,
        if(index == -1) {
            return true
        } else {
            // If the array is not yet ordered, remove the previous element from the index
            var arrayWithoutPreviousElement = sequence
            arrayWithoutPreviousElement.remove(at: index - 1)
            // And remove the next element from the index.
            var arrayWithoutNextElement = sequence
                arrayWithoutNextElement.remove(at: index + 1)
            // Check if the list orders after removing the previous or the next element from the index.
            return getUnorderedElementIndex(array: arrayWithoutPreviousElement) == -1
                || getUnorderedElementIndex(array: arrayWithoutNextElement) == -1
        }
    }
}

print(almostIncreasingSequence(sequence: [1, 4, 10, 4, 2])) // false
print(almostIncreasingSequence(sequence: [1, 2, 1, 2])) // false
print(almostIncreasingSequence(sequence: [1, 3, 2])) // true
print(almostIncreasingSequence(sequence: [1, 3, 2, 1])) // false
print(almostIncreasingSequence(sequence: [1, 3])) // true
print(almostIncreasingSequence(sequence: [1])) // true
print(almostIncreasingSequence(sequence: [])) // false